core/src/word_lock.rs

```diff
@@ -264,16 +266,45 @@ impl WordLock {
                     }
                 }
             } else {
-                (*queue_head).queue_tail.set(new_tail);
+                unsafe { (*queue_head).queue_tail.set(new_tail); }
                 self.state.fetch_and(!QUEUE_LOCKED_BIT, Ordering::Release);
             }
 
             // Finally, wake up the thread we removed from the queue. Note that
             // we don't need to worry about any races here since the thread is
             // guaranteed to be sleeping right now and we are the only one who
             // can wake it up.
-            (*queue_tail).parker.unpark_lock().unpark();
+            unsafe { (*queue_tail).parker.unpark_lock().unpark(); }
             break;
         }
     }
 }
+
+trait LockState {
+    fn is_locked(&self) -> bool;
+    fn is_queue_locked(&self) -> bool;
+    fn queue_head(&self) -> *const ThreadData;
+    fn set_queue_head(&self, thread_data: *const ThreadData) -> Self;
```

Comment:
This doesn't modify `self`, so maybe `with_queue_head` is a better name?

Reply:
That makes sense. Will also make every method consume `self` since it's just a `Copy` type anyway.